discovery/file/file.go

```diff
@@ -219,6 +289,9 @@ func (d *Discovery) refresh(ctx context.Context, ch chan<- []*config.TargetGroup
 	for f, n := range d.lastRefresh {
 		m, ok := ref[f]
 		if !ok || n > m {
+			level.Error(d.logger).Log("msg", "file_sd refresh found file that should be removed", "file", f)
+			// Remove from file sd timestamp collector.
+			fileSDTimeStamp.DeleteFile(f)
```

Comment:
This feels racy.

I think it'd be better to iterate over the active file SDs, each of which has its own state.

Reply:
this actually isn't needed to get the metric to not be reported if the file is removed, I don't quite understand what you mean by iterating over all active file SDs, that would mean we need to keep track of the timestamps in discovery.go wouldn't it?

Reply:
how about just clearing the map (or assigning a newly created one) to the timestamp collector at the start of refresh? since we always clear all files, and then compare the ones we knew about before to the ones we know about know for updating the target groups, we can just clear the map and then add timestamps for the ones that we're calling `readfile` on in `refresh`